lang/wasm.cc

```diff
@@ -178,7 +178,13 @@ struct from_val_visitor {
         if (ret_size == -1) {
             return bytes_opt{};
         }
-        bytes_opt ret = t.decompose(t.deserialize(bytes_view(reinterpret_cast<int8_t*>(data), ret_size)));
+        bytes_opt ret;
+        const counter_type_impl* counter = dynamic_cast<const counter_type_impl*>(&t);
+        if (counter) {
+            ret = long_type->decompose(t.deserialize(bytes_view(reinterpret_cast<int8_t*>(data), ret_size)));
+        } else {
+            ret = t.decompose(t.deserialize(bytes_view(reinterpret_cast<int8_t*>(data), ret_size)));
```

Comment:
Let's add a new operator() overlaod instead of a new special case.

Reply:
I'm afraid we will still need to handle the counter type in the `operator()(abstract_type)` because of https://github.com/scylladb/scylladb/blob/2653865b34d86ef338fa6f55fb29fe81675924bc/lang/wasm.cc#L276-L279

I can add it for the non-nullable case, but that will just be a copy paste of this `operator()`, unless we change our representation of counters in wasm